File: benchmarks/values/rational_harmonic.py
import time
import math

t0 = time.perf_counter()

num = 0
den = 1
for i in range(1, 3001):
    # num/den + 1/i = (num*i + den) / (den*i)
    num = num * i + den
    den = den * i
    # GCD reduce
    g = math.gcd(num, den)
    num = num // g
    den = den // g

digits = len(str(num))

t1 = time.perf_counter()
print(digits)
print(f"elapsed: {t1 - t0:.3f}s")

File: benchmarks/suite/rational_harmonic.go
package main

import (
	"fmt"
	"math/big"
	"time"
)

func main() {
	t0 := time.Now()

	num := big.NewInt(0)
	den := big.NewInt(1)
	g := new(big.Int)
	bi := new(big.Int)

	for i := int64(1); i <= 3000; i++ {
		bi.SetInt64(i)
		// num/den + 1/i = (num*i + den) / (den*i)
		num.Mul(num, bi)
		num.Add(num, den)
		den.Mul(den, bi)
		// GCD reduce
		g.GCD(nil, nil, num, den)
		num.Div(num, g)
		den.Div(den, g)
	}

	digits := len(num.String())

	elapsed := time.Since(t0)
	fmt.Println(digits)
	fmt.Printf("elapsed: %.3fs\n", elapsed.Seconds())
}
